csrc/scheduler/transpose.cpp

```diff
@@ -170,8 +170,12 @@ class DomainMap : public pointwise_utils::DomainMap {
     TensorView* result = nullptr;
     int64_t max_dims = -1;
     for (auto tv : group) {
-      if (isValidReference(tv)) {
-        int64_t dims = (int64_t)pointwise_utils::nRootDims(tv);
+      // since transpose scheduler have different set of reference, we skip IDs
```

Comment:
> Each of the reference tensors is supposed to pass isValidReferece, so I think each reference should be mapped with all the input tensors. 

What I was trying to say is that, this is what transpose scheduler is doing today, and I don't think that's the right thing to do neither.

If you look at how the scheduler is propagating the parallelization strategy.
For [group 2](https://github.com/NVIDIA/Fuser/blob/608868f557464282480640564e82b888833ae804/csrc/scheduler/transpose.cpp#L1254-L1300) and for [group 1](https://github.com/NVIDIA/Fuser/blob/608868f557464282480640564e82b888833ae804/csrc/scheduler/transpose.cpp#L1331-L1388). The propagation is limited to other TVs in its own group. It's not supposed to be propagated to all inputs and outputs.

I think we took a short-cut in the first place to re-use the `pointwise_utils::isValidReference` when we check for legitimacy of each reference.

TL;DR, if we apply the same check for all outputs, we are going to fail some CI tests that checks the transpose scheduler is taking the fusion.
I think we need to have a follow up PR that adds a `transpose_utils::isValidReference` instead of re-use `pointwise_utils::isValidReference`

Reply:
I'm not sure why transposing would create disconnected tensors. Can you show an example?

Reply:
In this example

```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id38(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[1024, 128], contiguity=[True, True], dtype=DataType.Float, is_cpu=False, stride_order=[1, 0])
    T1 = fd.define_tensor(shape=[1024, 128], contiguity=[True, True], dtype=DataType.Float, is_cpu=False, stride_order=[0, 1])

    T2 = fd.ops.broadcast_in_dim(T1, shape=[1, 1024, 128], broadcast_dims=[1, 2])

    T3 = fd.ops.broadcast_in_dim(T0, shape=[1, 1024, 128], broadcast_dims=[1, 2])
    T4 = fd.ops.broadcast_in_dim(T3, shape=[32, 1024, 128], broadcast_dims=[0, 1, 2])

    T5 = fd.ops.add(T4, T2)
    T6 = fd.ops.reshape(T5, new_shape=[32*1024, 128])
    fd.add_output(T6)

with FusionDefinition() as fd:
    nvfuser_fusion_id38(fd)

inputs = [
    torch.testing.make_tensor((1024, 128), dtype=torch.float32, device='cuda:0'),
    torch.testing.make_tensor((1024*128), dtype=torch.float32, device='cuda:0').as_strided((1024, 128), (1, 1024)),
]
o = fd.execute(inputs)

def foo(inputs):
    T2 = inputs[1].unsqueeze(0)
    T3 = inputs[0].expand((32, 1024, 128))
    T4 = T2 + T3
    return T4.reshape(32*1024, 128)

o_ref = foo(inputs)

assert o[0].allclose(o_ref)
```

```
g{(transpose)
group id: 0
inputs:
  T0_g_float[iS0{1024}, iS1{128}] float
  T1_g_float[iS2{1024}, iS3{128}] float
outputs:
  T9_g_float[iS30{32768}rf, iS27{128}] float


T4_l_float[bS10{1}, iS11{1024}, iS12{128}]
   = broadcast( T0_g_float[iS0{1024}, iS1{128}] )
(2)
T7_l_float[bS19{1 ex 32}, iS20{1024}, iS21{128}] = expand( T4_l_float[bS10{1}, iS11{1024}, iS12{128}], {32, 1024, 128} )
(29)
T2_l_float[bS4{1}, iS5{1024}, iS6{128}]
   = broadcast( T1_g_float[iS2{1024}, iS3{128}] )
(0)
T8_l_float[bS22{1 ex 32}, iS23{1024}, iS24{128}]
   = T7_l_float[bS19{1 ex 32}, iS20{1024}, iS21{128}]
   + T2_l_float[bS4{1}, iS5{1024}, iS6{128}];
(27)
T9_g_float[iS30{32768}rf, iS27{128}] = view( T8_l_float[bS22{1 ex 32}, iS23{1024}, iS24{128}] )
(8)
}

} //Segmented_Fusion


===== Transpose Stats ========
inputs: T0_g_float[iS0{1024}, iS1{128}], T1_g_float[iS2{1024}, iS3{128}]
outputs: T9_g_float[iS30{32768}rf, iS27{128}]
shape: 32768 128
num_elems: 4194304
n_io_tensors: 3
max_io_dtype_size: 4
group 1: T9_g_float[iS30{32768}rf, iS27{128}], T0_g_float[iS0{1024}, iS1{128}]
reference1: T9_g_float[iS30{32768}rf, iS27{128}]
inner_most_id1 position: 1 (in reference 1)
group 2: T1_g_float[iS2{1024}, iS3{128}]
reference2: T1_g_float[iS2{1024}, iS3{128}]
inner_most_id2 position: 0 (in reference 1)
```

Here reference2 : T1_g_float[iS2{1024}, iS3{128}] doesn't cover all the source IDs from output `T9`, but transpose scheduler only propagate/parallelize reference2 within its own group, so it's still a valid reference.

Ifw e don't skip the check here, we'll end up not using pointwise scheduler and loose part of the vectorization..

Reply:
I see. It seems this would miss a case where a reference tensor indeed has an output like `T9` in the same group. If so, we should not just always skip the check even for the transpose scheduler, should we?

Reply:
indeed. It's not safe to blindly skip it. We want the check to be consistent of how transformation is propagated.

I have a comment in the transpose scheduler when we are passing `false` to skip the check and how it's not correct.
Transpose scheduler is a bit strange, since it also does a global scheduling at Step 2, but only to a subset of IDs on reference1. So the right check to apply is a bit blurry.

I feel I needed to look at it further to figure out what is the right check for transpose scheduler and I'm not ready to commit to that in this PR.

Reply:
I see. I thought you're saying this would be less than ideal but still acceptable, but looks like this is a real bug. I agree that fixing the issue doesn't need to be done in this PR, but I think we should take the safer option, i.e., don't skip the check even with the transpose scheduler. Failing to use a vectorized transpose scheduler is a bummer but is better than failing to run.

Reply:
Sounds fair, I'll go with that route and temporarily disable some CI tests. I'll make sure I open an issue to track the disabled tests and restore them when we fix the validation.

Reply:
Opened issue #3570. I'll have it in the code comment as well.